File: leetcode/medium/684_findRedundantConnection.py
from typing import List
import json
from collections import deque

class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        parents = [i for i in range(n+1)]
        rank = [1 for i in range(n+1)]

        def find(id):
            if parents[id] != id:
                parents[id] = find(parents[id])
            return parents[id]

        def union(s, e):
            p1 = find(s)
            p2 = find(e)
            if p1 == p2:
                return True

            if rank[p2] > rank[p1]:
                rank[p2] += rank[p1]
                parents[p1] = p2
            else:
                rank[p1] += rank[p2]
                parents[p2] = p1

            return False

        for s, e in edges:
            if union(s,e):
                return [s, e]

        return []

def test ():
    params = [
        {
            'input': [[1,2],[1,3],[2,3]],
            'output': [2,3],
        },
        {
            'input': [[1,2],[2,3],[3,4],[1,4],[1,5]],
            'output': [1,4],
        },
    ]
    solution = Solution()

    for param in params:
        result = solution.findRedundantConnection(param['input'])
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if json.dumps(result) == json.dumps(param['output']) else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()

File: leetcode/medium/684_findRedundantConnection.ts
function findRedundantConnection(edges: number[][]): number[] {
    const parents: Map<number, number> = new Map();
    const childrenCount: Map<number, number> = new Map();

    const getParent = (num: number): number => {
        if (!parents.has(num)) {
            parents.set(num, num);
            childrenCount.set(num, 1);
            return num;
        }

        const parent: number = parents.get(num);
        if (parent === num) {
            return num;
        }

        const grandParent: number = getParent(parent);

        if (parent !== grandParent) {
            parents.set(num, grandParent);
        }

        return grandParent;
    };
    const setParent = (num1: number, num2: number): void => {
        const parent1: number = getParent(num1);
        const parent2: number = getParent(num2);

        if (parent1 === parent2) {
            return;
        }

        const count1: number = childrenCount.get(parent1);
        const count2: number = childrenCount.get(parent2);

        if (count1 < count2) {
            parents.set(parent1, parent2);
            childrenCount.set(parent2, count1 + count2);
        } else {
            parents.set(parent2, parent1);
            childrenCount.set(parent1, count1 + count2);
        }
    };

    let res: number[] = [];

    for (const [s, e] of edges) {
        const pS: number = getParent(s);
        const pE: number = getParent(e);

        if (pS === pE) {
            res = [s, e];
        }

        setParent(s, e);
    }

    return res;
};

const test = () => {
    const params = [
        {
            input: {
                edges: [[1,2],[1,3],[2,3]],
            },
            output: [2,3],
        },
        {
            input: {
                edges: [[1,2],[2,3],[3,4],[1,4],[1,5]],
            },
            output: [1,4],
        },
    ];

    params.forEach(({input, output}) => {
        const { edges } = input;
        const result = findRedundantConnection(edges);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();

